keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -1656,30 +1662,26 @@ gets updated now if the field doesn't
 match the current list already. If no PodScheduling object exists yet,
 it gets created.
 
-When there is a scheduler extender configured for the claim, creating and
-updating PodScheduling objects gets skipped because the scheduler
-extender handles filtering.
-
 #### Reserve
 
 A node has been chosen for the Pod.
 
 If using delayed allocation and the resource has not been allocated yet,
-the SelectedNode field of the ResourceClaim
+the PodSchedulingSpec.SelectedNode field
 gets set here and the scheduling attempt gets stopped for now. It will be
-retried when the ResourceClaim status changes.
+retried when the ResourceClaim or PodScheduling statuses change.
 
 If all resources have been allocated already,
 the scheduler adds the Pod to the `claim.status.reservedFor` field of its ResourceClaims to ensure that
 no-one else gets to use those.
 
 If some resources are not allocated yet or reserving an allocated resource
 fails, the scheduling attempt needs to be aborted and retried at a later time
-or when the ResourceClaims change.
+or when the statuses change (same as above).
 
 #### Unreserve
 
-The scheduler removes the Pod from the ReservedFor field because it cannot be scheduled after
+The scheduler removes the Pod from the ResourceClaimStatus.ReservedFor field because it cannot be scheduled after
```

Comment:
Maybe remove the SelectedNode instead?

Reply:
Doing only that prevents other pods from using the claim. If one pod is potentially stuck while waiting for some other resource, it shouldn't hold onto claims that might be used by some other pod in the meantime.

This leads to a bigger question of how long those claims should be kept in the allocated state. The underlying resources might be needed for some other claim. Let's be optimistic and assume that most of the time, allocated claims will be put to good use quickly, okay?

We can still tune this once people have some real-world experience with this.

Reply:
Hmm, after re-reading what we said above about triggering deallocating by clearing the `reservedFor` field it becomes clear that the scheduler must keep the claim reserved here. Otherwise we'll have a lot of thrashing for a pod with several claims:
- several drivers start allocating for the selectedNode
- most allocations succeed, one doesn't
- `reservedFor` gets cleared, `selectedNode` unset
- all drivers start deallocating
- once that is done, allocation is attempted anew for all claims

I think the same argument about being optimistic also applies to `reservedFor`: let's keep it set as you suggested and hope that the situation gets resolved quickly or (even better) doesn't occur often.

Reply:
After sleeping over this I realized that this approach with keeping claims reserved has a drawback.

Suppose there are two separate claims and two pods both referencing them. Both pods get scheduled concurrently. One pod manages to allocate one claim and reserves it, the other pod does the same for the other claim. Now both pods are stuck because they only have one out of two claims.

This can happen both for claims with delayed allocation and with immediate allocation.

To solve this, we would need additional logic somewhere that looks at multiple different pods to detect such a deadlock and then resolve it. I don't know where to put that.

Perhaps the original proposal with giving up reservations and deallocation triggered explicitly is better after all? It relies on chance and repeated attempts to get one of the two pods scheduled, i.e. there's no coordination either, but it's not needed because eventually one pod should succeed.

Reply:
I've added this to the KEP as explanation why Unreserve must remove the pod.

Reply:
It's not just the risk of deadlocks. Consider claims X and Y and pods A, referencing X and Y, and pod B, referencing X, and pod C, referencing Y. If A reserves X and C reserves Y, then only pod C can run. A waits for C to finish and release Y, B waits for A.

If A releases X right after the failed scheduling attempt, then both B and C can run.